### ISSUE TITLE
Can't run trainer/train.py well

### ISSUE THREAD

**Issue by fxf000** [google-research/deeplab2#36]
1, Stopped halfway the trainer/train.py because the learning didn't progress at all.
>2021-07-20 06:31:51.027365: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-07-20 06:31:51.032914: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/lib64-nvidia
2021-07-20 06:31:51.033924: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1835] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
I0720 06:31:51.034065 140350081660800 train.py:65] Reading the config file.
I0720 06:31:51.036973 140350081660800 train.py:69] Starting the experiment.
2021-07-20 06:31:51.037464: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0720 06:31:51.038176 140350081660800 train_lib.py:105] Using strategy <class 'tensorflow.python.distribute.one_device_strategy.OneDeviceStrategy'> with 1 replicas
I0720 06:31:51.042598 140350081660800 deeplab.py:57] Synchronized Batchnorm is used.
I0720 06:31:51.043589 140350081660800 axial_resnet_instances.py:144] Axial-ResNet final config: {'num_blocks': [3, 4, 6, 3], 'backbone_layer_multiplier': 1.0, 'width_multiplier': 1.0, 'stem_width_multiplier': 1.0, 'output_stride': 16, 'classification_mode': False, 'backbone_type': 'resnet_beta', 'use_axial_beyond_stride': 16, 'backbone_use_transformer_beyond_stride': 32, 'extra_decoder_use_transformer_beyond_stride': 32, 'backbone_decoder_num_stacks': 0, 'backbone_decoder_blocks_per_stage': 1, 'extra_decoder_num_stacks': 0, 'extra_decoder_blocks_per_stage': 1, 'max_num_mask_slots': 128, 'num_mask_slots': 128, 'memory_channels': 256, 'base_transformer_expansion': 1.0, 'global_feed_forward_network_channels': 256, 'high_resolution_output_stride': 4, 'activation': 'relu', 'block_group_config': {'attention_bottleneck_expansion': 2, 'drop_path_keep_prob': 0.800000011920929, 'drop_path_beyond_stride': 16, 'drop_path_schedule': 'linear', 'positional_encoding_type': None, 'use_global_beyond_stride': 0, 'use_sac_beyond_stride': -1, 'use_squeeze_and_excite': False, 'conv_use_recompute_grad': False, 'axial_use_recompute_grad': True, 'recompute_within_stride': 0, 'transformer_use_recompute_grad': False, 'axial_layer_config': {'query_shape': (129, 129), 'key_expansion': 1, 'value_expansion': 2, 'memory_flange': (32, 32), 'double_global_attention': False, 'num_heads': 8, 'use_query_rpe_similarity': True, 'use_key_rpe_similarity': True, 'use_content_similarity': True, 'retrieve_value_rpe': True, 'retrieve_value_content': True, 'initialization_std_for_query_key_rpe': 1.0, 'initialization_std_for_value_rpe': 1.0, 'self_attention_activation': 'softmax'}, 'dual_path_transformer_layer_config': {'num_heads': 8, 'bottleneck_expansion': 2, 'key_expansion': 1, 'value_expansion': 2, 'feed_forward_network_channels': 2048, 'use_memory_self_attention': True, 'use_pixel2memory_feedback_attention': True, 'transformer_activation': 'softmax'}}, 'bn_layer': functools.partial(<class 'keras.layers.normalization.batch_normalization.SyncBatchNormalization'>, momentum=0.9900000095367432, epsilon=0.0010000000474974513), 'conv_kernel_weight_decay': 0.0}
I0720 06:31:51.322444 140350081660800 deeplab.py:96] Setting pooling size to (65, 65)
I0720 06:31:51.322864 140350081660800 aspp.py:135] Global average pooling in the ASPP pooling layer was replaced with tiled average pooling using the provided pool_size. Please make sure this behavior is intended.
I0720 06:31:57.261267 140350081660800 controller.py:391] restoring or initializing model...
restoring or initializing model...
I0720 06:31:57.299474 140350081660800 controller.py:395] restored model from ../drive/MyDrive/checkpoints/ImageNet_pretrained_checkpoints/max_deeplab_s_os16_res1025_100k_coco_train/ckpt-0.
restored model from ../drive/MyDrive/checkpoints/ImageNet_pretrained_checkpoints/max_deeplab_s_os16_res1025_100k_coco_train/ckpt-0.
I0720 06:31:57.299665 140350081660800 controller.py:217] restored from checkpoint: ../drive/MyDrive/checkpoints/ImageNet_pretrained_checkpoints/max_deeplab_s_os16_res1025_100k_coco_train/ckpt-0
restored from checkpoint: ../drive/MyDrive/checkpoints/ImageNet_pretrained_checkpoints/max_deeplab_s_os16_res1025_100k_coco_train/ckpt-0
I0720 06:31:58.465277 140350081660800 api.py:446] Eval with scales ListWrapper([1.0])
I0720 06:31:59.732715 140350081660800 api.py:446] Global average pooling in the ASPP pooling layer was replaced with tiled average pooling using the provided pool_size. Please make sure this behavior is intended.
I0720 06:31:59.757121 140350081660800 api.py:446] Eval scale 1.0; setting pooling size to [65, 65]
I0720 06:32:12.421342 140350081660800 api.py:446] Global average pooling in the ASPP pooling layer was replaced with tiled average pooling using the provided pool_size. Please make sure this behavior is intended.
I0720 06:32:13.528391 140350081660800 controller.py:236] train | step:      0 | training until step 3...
train | step:      0 | training until step 3...
2021-07-20 06:32:13.646223: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
2021-07-20 06:32:14.011432: W tensorflow/core/framework/dataset.cc:679] Input of GeneratorDatasetOp::Dataset will not be optimized because the dataset does not implement the AsGraphDefInternal() method needed to apply optimizations.
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/util/dispatch.py:206: calling foldl_v2 (from tensorflow.python.ops.functional_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.foldl(fn, elems, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.foldl(fn, elems))
W0720 06:32:25.062542 140350081660800 deprecation.py:616] From /usr/local/lib/python3.7/dist-packages/tensorflow/python/util/dispatch.py:206: calling foldl_v2 (from tensorflow.python.ops.functional_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.foldl(fn, elems, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.foldl(fn, elems))
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/impl/api.py:463: calling while_loop_v2 (from tensorflow.python.ops.control_flow_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.while_loop(c, b, vars, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.while_loop(c, b, vars))
W0720 06:32:25.398839 140350081660800 deprecation.py:616] From /usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/impl/api.py:463: calling while_loop_v2 (from tensorflow.python.ops.control_flow_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.while_loop(c, b, vars, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.while_loop(c, b, vars))
^C

2, Command tried
>!python ./deeplab2/trainer/train.py \
    --config_file=./deeplab2/configs/coco/max_deeplab/max_deeplab_s_os16_res1025_100k.textproto \
    --mode='train' \
    --model_dir=../drive/MyDrive/checkpoints/ImageNet_pretrained_checkpoints/ \
    --num_gpus=1

3, Env
>On google colab; 
	tf.__version__ = 2.6.0-rc1

I think the GPU is not working prperly.
I'd appreciate it if anyone could tell me how to solve it.

**Comment by ClaireXie** [comment]
Hi fxf000,

From the log, I saw "Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory;". Can you make sure you have libcusolver installed? If not, you can install it via `sudo apt-get install libcusolver11` and try running training again.

**Comment by fxf000** [comment]
Hi ClaireXie,

Thank you for answering.
I tried the command below, but I coudn’t use the GPU again.

· command
>!sudo ln -s /usr/local/cuda-11.0/targets/x86_64-linux/lib/libcusolver.so.10 /usr/local/cuda-11.0/targets/x86_64-linux/lib/libcusolver.so.11

·error(same)
>2021-07-23 05:41:07.415835: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-07-23 05:41:07.729413: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/lib64-nvidia
2021-07-23 05:41:07.750896: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1835] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.

I thought that `LD_LIBRARY_PATH` was not set correctly, so I wrote `/etc/profile` to `LD_LIBRARY_PATH=$LD_LIBRARY_PATH:../../usr/local/cuda-11.1/targets/x86_64-linux/lib export LD_LIBRARY_PATH` and did `!source /etc/profile`, but it didn't work.

**Comment by csrhddlam** [comment]
Hi fxf000,

Thank you for your interest in our work and the effort at running the code.
There seems to be around two months since the last update. Are you able to load the 'libcusolver.so.11' now?

**Comment by aquariusjay** [comment]
Closing the issue since it is not active for a while.
Please feel free to reopen the issue if needed.
